app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt

```diff
@@ -0,0 +1,556 @@
+                    routerOrigin: RouterOrigin
+                ) {
+                    setRoute(routes)
+                }
+
+                override fun onFailure(
+                    reasons: List<RouterFailure>,
+                    routeOptions: RouteOptions
+                ) {
+                    // no impl
+                }
+
+                override fun onCanceled(routeOptions: RouteOptions, routerOrigin: RouterOrigin) {
+                    // no impl
+                }
+            }
+        )
+    }
+
+    private fun setRoute(routes: List<DirectionsRoute>) {
+        // set routes, where the first route in the list is the primary route that
+        // will be used for active guidance
+        mapboxNavigation.setRoutes(routes)
+
+        // move the camera to overview when new route is available
+        navigationCamera.requestNavigationCameraToOverview()
+
+        binding.playRouteButton.show()
+        binding.playRouteButton.setOnClickListener {
+            // simulate the current route
+            startSimulation(routes.first())
+        }
+    }
+
+    private fun startSimulation(route: DirectionsRoute) {
+        binding.playRouteButton.hide()
+        mapboxReplayer.run {
+            stop()
+            clearEvents()
+            val replayEvents = ReplayRouteMapper().mapDirectionsRouteGeometry(route)
+            pushEvents(replayEvents)
+            seekTo(replayEvents.first())
+            play()
+        }
+    }
+
+    private fun stopSimulation() {
+        mapboxReplayer.run {
+            stop()
+            clearEvents()
+        }
+    }
+}
+
+class WaypointsSet {
+
+    private val waypoints = mutableListOf<Waypoint>()
+
+    val isEmpty get() = waypoints.isEmpty()
+
+    fun addNamed(point: Point, name: String) {
+        waypoints.add(Waypoint(point, WaypointType.Named(name)))
+    }
+
+    fun addRegular(point: Point) {
+        waypoints.add(Waypoint(point, WaypointType.Regular))
+    }
+
+    fun addSilent(point: Point) {
+        waypoints.add(Waypoint(point, WaypointType.Silent))
+    }
+
+    fun clear() {
+        waypoints.clear()
+    }
+
+    fun waypointsIndices(): List<Int> {
+        return waypoints.mapIndexedNotNull { index, waypoint ->
+            if (waypoint.type == WaypointsSet.WaypointType.Silent) {
+                null
+            } else index
+        }
+    }
+
+    fun waypointsNames(): List<String> = waypoints.filter { it.type != WaypointType.Silent }.map {
+        when (it.type) {
+            is WaypointType.Named -> it.type.name
+            else -> ""
+        }
+    }
+
+    fun coordinatesList(): List<Point> {
+        return waypoints.map { it.point }
+    }
+
+    private sealed class WaypointType {
+        data class Named(val name: String) : WaypointType()
+        object Regular : WaypointType()
+        object Silent : WaypointType()
+    }
+
+    private data class Waypoint(val point: Point, val type: WaypointType)
+}
```

Comment:
Where do you usually put helper classes? Logic of converting waypoints to params is too complex to inline it in activity.

Reply:
Okay, I'm leaving it here if there are no strong opinions 🙂

Reply:
Yeah the old examples, found the need for a [helper class too](https://github.com/mapbox/mapbox-navigation-android/blob/main/examples/src/main/java/com/mapbox/navigation/examples/core/waypoints/WaypointsController.kt). IMO it's fine to have in here.

But also, the fact we need a helper class makes me think we could design a class to handle this in core-sdk.